Use offsetX/offsetY to compute mouse position in default raycaster

axelboc opened the pull request:
Fix #350 

The canvas-relative cursor coordinates given to the default raycaster used to be computed based on the position of the cursor in relation to the viewport (`clientX`/`clientY`) and the position of the canvas (`left`/`top`) provided by `useMeasure` (via `ResizeContainer.tsx`).

When the canvas moves on the screen without being resized or without the viewport being scrolled or resized, `useMeasure` has no way of knowing it, so `prepareRay` used to rely on potentially outdated canvas positions.

With this fix, `prepareRay` now uses the cursor's `offsetX` and `offsetY` positions, which are already canvas-relative coordinates, so the position of the canvas in the viewport no longer matters.

---

If I'm not 100% sure, but I think this fix means that raycasting will now work regardless of scroll. So it may be possible to disable scroll detection in `useMeasure` by setting the `scroll` configuration option to `false` by default.

axelboc commented:
I've seen other uses of `clientX` and `clientY` in the source code:

- in event listeners in some of the demos, and
- for the computation of `delta` in `calculateDistance`.

Using `offsetX` and `offsetY` would work as well, of course, but it's not necessary, so I left these cases alone.

axelboc commented:
Any news on this? I could really use the fix 🙏 😄

drcmda commented:
sorry for the delay, @axelboc im a little nervous about dom events and such. is this safe to merge?

axelboc commented:
I'm pretty confident it is. Can someone give a second opinion?

drcmda commented:
and the other places where it's still using clientX/Y? could we change them to offsets as well? i have pretty much given up understanding dom events at this point but it would maintenance easier if there's at least some unity.

axelboc commented:
Sorry for the delay in replying, I had moved to another project for a while. I had a look at the other uses of `clientX/Y` this morning:

- Delta computation in `src/canvas.tsx` (`state.initialClick`, `calculateDistance`) => **replaced with `offsetX/Y`** ✔️
- `onMouseMove` handlers in demos => **not worth replacing** because the code is simpler with `clientX/Y` (with `offsetX/Y`, we need to know the canvas' size, so the computation can't be done in the `onMouseMove` event listener that is set on the `Canvas` element).

I've also rebased the branch and fixed a couple of linting issues in the `Physics` demo.

axelboc commented:
I've rebased on the latest v5 changes. This is ready to merge.

drcmda commented:
sorry it took so long, its just that im scared to even touch browser events since it has failed so often before. lets just try it. i also disabled scroll detection.
